src/tasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,365 @@
+        var resourceSymbols = new StringBuilder();
+        // Generate source file(s) containing each resource's byte data and size
+        if (remainingDestinationFilesToBundle.Length > 0)
+        {
+            int allowedParallelism = Math.Max(Math.Min(remainingDestinationFilesToBundle.Length, Environment.ProcessorCount), 1);
+            if (BuildEngine is IBuildEngine9 be9)
+                allowedParallelism = be9.RequestCores(allowedParallelism);
+            if (!string.IsNullOrEmpty(CombinedResourceSource))
+                allowedParallelism = 1;
+
+            Parallel.For(0, remainingDestinationFilesToBundle.Length, new ParallelOptions { MaxDegreeOfParallelism = allowedParallelism, CancellationToken = BuildTaskCancelled.Token }, (i, state) =>
+            {
+                var group = remainingDestinationFilesToBundle[i];
+
+                // Since the object filenames include a content hash, we can pick an arbitrary ITaskItem from each group,
+                // since we know each group's ITaskItems all contain the same binary data
+                var contentSourceFile = group.First();
+
+                var outputFile = group.Key;
+                if (!string.IsNullOrEmpty(CombinedResourceSource))
+                    outputFile = Path.Combine(OutputDirectory, CombinedResourceSource);
+                var inputFile = contentSourceFile.ItemSpec;
+                bool shouldAddHeader = false;
+                lock (resourceSourcesGenerated)
+                {
+                    shouldAddHeader = !resourceSourcesGenerated.Contains(outputFile);
+                }
+
+                if (verbose)
+                {
+                    var registeredName = contentSourceFile.GetMetadata("RegisteredName");
+                    if(string.IsNullOrEmpty(registeredName))
+                    {
+                        registeredName = Path.GetFileName(inputFile);
+                    }
+                    var count = Interlocked.Increment(ref verboseCount);
+                    Log.LogMessage(MessageImportance.Low, "{0}/{1} Bundling {2} ...", count, remainingDestinationFilesToBundle.Length, registeredName);
+                }
+
+                Log.LogMessage(MessageImportance.Low, "Bundling {0} into {1}", inputFile, outputFile);
+                var symbolName = ToSafeSymbolName(group.Key);
+                if (!Emit(outputFile, (codeStream) => {
+                    using var inputStream = File.OpenRead(inputFile);
+                    BundleFileToCSource(symbolName, shouldAddHeader, inputStream, codeStream);
+                }))
+                {
+                    state.Stop();
+                }
+                lock (resourceSourcesGenerated)
+                {
+                    resourceSourcesGenerated.Add(outputFile);
+                }
+                resourceSymbols.AppendLine($"extern uint8_t {symbolName}_data[];");
+                resourceSymbols.AppendLine($"extern const uint32_t {symbolName}_len;");
```

Comment:
I think the len symbol should be named:

extern const uint32_t {symbolName}_data_len;